### PR TITLE
Better barcode support

### PR DESCRIPTION
**Pull request by RafDevX**
(not tested)

### CODE REVIEW

File: src/grocy.rs [hunk 1]
```diff
@@ -297,3 +322,39 @@ fn prompt_location(grocy_state: &GrocyState, default_location: u32) -> Result<Lo
...
+                Some(Box::new(MultiSelect::new(&msg, options.unwrap())))
+            }
+            _ => None, // unsupported type
+        };
+
+        if let Some(prompt) = prompt {
+            if field.input_required {
+                values.insert(field.name, prompt.prompt()?);
+            } else if let Some(value) = prompt.prompt_skippable()? {
+                values.insert(field.name, value);
```
Comment (RafDevX) [inline src/grocy.rs]: This doesn't work... 
```
error[E0161]: cannot move a value of type `dyn dynprompt::Prompt`
   --> src/grocy.rs:352:43
    |
352 |                 values.insert(field.name, prompt.prompt()?);
    |                                           ^^^^^^ the size of `dyn dynprompt::Prompt` cannot be statically determined

```

Comment (diogotcorreia) [reply src/grocy.rs]: https://stackoverflow.com/questions/76035679/cannot-move-a-value-of-dyn-type

Use generics instead.

```diff
diff --git a/src/grocy.rs b/src/grocy.rs
index 9cc8d2b..6d2904f 100644
--- a/src/grocy.rs
+++ b/src/grocy.rs
@@ -4,7 +4,7 @@ use std::{collections::HashMap, fmt::Display};
 use anyhow::Result;
 use chrono::{Duration, NaiveDate};
 use grocy::{
-    structs::{Location, ProductDetails},
+    structs::{Location, ProductDetails, UserField},
     GrocyApi,
 };
 use inquire::{Confirm, CustomType, DateSelect, MultiSelect, Select, Text};
@@ -334,27 +334,31 @@ fn prompt_barcode_userfields(grocy_api: &GrocyApi) -> Result<HashMap<String, Str
             .as_deref()
             .map(|config| config.lines().map(|l| l.to_owned()).collect());
 
-        let prompt: Option<Box<dyn Prompt>> = match field.r#type.as_str() {
-            "checkbox" => Some(Box::new(Confirm::new(&msg))),
-            "text-single-line" => Some(Box::new(Text::new(&msg))),
-            "date" => Some(Box::new(DateSelect::new(&msg))),
+        let value: Option<String> = match field.r#type.as_str() {
+            "checkbox" => prompt_userfield(Confirm::new(&msg), &field)?,
+            "text-single-line" => prompt_userfield(Text::new(&msg), &field)?,
+            "date" => prompt_userfield(DateSelect::new(&msg), &field)?,
             "preset-list" if options.is_some() => {
-                Some(Box::new(Select::new(&msg, options.unwrap())))
+                prompt_userfield(Select::new(&msg, options.unwrap()), &field)?
             }
             "preset-checklist" if options.is_some() => {
-                Some(Box::new(MultiSelect::new(&msg, options.unwrap())))
+                prompt_userfield(MultiSelect::new(&msg, options.unwrap()), &field)?
             }
             _ => None, // unsupported type
         };
 
-        if let Some(prompt) = prompt {
-            if field.input_required {
-                values.insert(field.name, prompt.prompt()?);
-            } else if let Some(value) = prompt.prompt_skippable()? {
-                values.insert(field.name, value);
-            }
+        if let Some(value) = value {
+            values.insert(field.name, value);
         }
     }
 
     Ok(values)
 }
+
+fn prompt_userfield<T: Prompt>(prompt: T, field: &UserField) -> Result<Option<String>> {
+    if field.input_required {
+        Ok(Some(prompt.prompt()?))
+    } else {
+        Ok(prompt.prompt_skippable()?)
+    }
+}
```

File: grocy/src/structs.rs [hunk 2]
```diff
@@ -52,19 +53,38 @@ pub struct QuantityUnit {
...
+
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct AddProductBarcodePayload<'a> {
     pub product_id: u32,
     pub barcode: &'a str,
     pub amount: Option<f64>,
     pub qu_id: Option<u32>,
+    pub shopping_location_id: u32,
+    pub note: Option<&'a str>,
+    pub userfields: HashMap<String, String>,
```
Comment (diogotcorreia) [inline grocy/src/structs.rs]: This is a separate request (please only do it if there are user fields)

**Review by diogotcorreia — CHANGES_REQUESTED** [review]

File: grocy/src/structs.rs [hunk 3]
```diff
@@ -118,3 +120,13 @@ pub struct PurchaseProductPayload<'a> {
 pub struct Transaction {
     pub id: u32,
 }
+
+// e.g., for when deserializing -1 into an Option<i32>
```
Comment (diogotcorreia) [inline grocy/src/structs.rs]:
```suggestion
// e.g., for when deserializing -1 into an Option<u32>
```

File: src/grocy.rs [hunk 4]
```diff
@@ -319,3 +322,36 @@ fn prompt_location(grocy_state: &GrocyState, default_location: u32) -> Result<Lo
 
     Ok(location)
 }
+
+fn prompt_barcode_userfields(grocy_api: &GrocyApi) -> Result<HashMap<String, String>> {
+    let userfields = grocy_api.get_barcode_userfields()?;
+    let mut values = HashMap::new();
+
+    for field in userfields {
+        let msg = format!("{} for this barcode:", field.caption);
```
Comment (diogotcorreia) [inline src/grocy.rs]: Include something in the message that tells the user the field is optional/mandatory (it's not obvious from the prompt if it's optional or not)

File: src/grocy.rs [hunk 5]
```diff
@@ -319,3 +322,36 @@ fn prompt_location(grocy_state: &GrocyState, default_location: u32) -> Result<Lo
...
+        let options = field
+            .config
+            .as_deref()
+            .map(|config| config.lines().map(|l| l.to_owned()).collect());
+
+        let value: Option<String> = match field.r#type.as_str() {
+            "checkbox" => dynprompt::prompt(Confirm::new(&msg), field.input_required)?,
```
Comment (diogotcorreia) [inline src/grocy.rs]: image
The prompt doesn't say anything about it being `yes` or `no`, only after submitting an invalid answer does it tell you that.
Perhaps add a help message/placeholder saying it's a yes/no field.

**Review by diogotcorreia — CHANGES_REQUESTED** [review]

**Review by diogotcorreia — APPROVED** [review]

**Pull request merged**